wasm.md

```diff
@@ -224,6 +224,33 @@ All of the following opcodes are liftings of the K builtin operators using the h
     rule <k> ITYPE . ge_s I1 I2 => < ITYPE > #bool(#signed(ITYPE, I1) >=Int #signed(ITYPE, I2)) ... </k>
 ```
 
+### Conversion Operations
+
+These operators convert constant elements at the top of the stack to another type. The target type is before the `.`, and the source type is after the `_`.
+
+Wrapping cuts of the 32 most significant bits of an `i64` value.
+
+```k
+    syntax Instr ::= "i32" "." "wrap_i64"
+ // -------------------------------------
+ rule <k> i32 . wrap_i64 => . ... </k>
+      <stack> (< i64 > I  => #chop(< i32 > I)) : STACK </stack>
+```
+
+Extension turns an `i32` type value into the corresponding `i64` type value.
+
+```k
+    syntax Instr ::= "i64" "." "extend_i32_u"
+ // -----------------------------------------
+    rule <k> i64 . extend_i32_u => . ... </k>
+         <stack> (< i32 > N => < i64 > N) : STACK </stack>
+
+    syntax Instr ::= "i64" "." "extend_i32_s"
+ // -----------------------------------------
+    rule <k> i64 . extend_i32_s => . ... </k>
+         <stack> (< i32 > N => < i64 > #unsigned(i64, #signed(i32, N))) : STACK </stack>
+```
```

Comment:
Hmmmmm, this all looks correct, but it bothers me that we can't use the `UnStackOp` machinery for loading arguments and type-checking them. I guess it's because the type before the operator is about the *return* type, and when we're type-checking opcodes we are treating it like the *input* type. This may be worth discussing at the next meeting.

Mostly I just want to make sure you have thought about this, so let me know that you have and I'll mark this as resolved.

Reply:
Sorry for the brain-dump, but I dove into this question for a this morning to try to come up with alternative solutions. Happy to discuss this further on Tuesday.

UnOp
-------
I came up with 2 other solutions which use the IUnOp machinery.
 ```k
    syntax IUnOp ::= "wrap_i64"
 // ---------------------------
 rule <k> i32 . wrap_i64 => i64 . wrap_i64 ... </k>
 rule <k> i64 . wrap_i64 I1 => #chop(< i32 > I1) ... </k>
```
and
```k
    syntax IUnOp ::= "wrap_i64" | #wrap ( IValType )
 // ------------------------------------------------
 rule <k> i32 . wrap_i64 => i64 . #wrap ( i32 ) ... </k>
 rule <k> _ . #wrap ( ITYPE ) I1 => #chop( < ITYPE > I1) ... </k>
```

Both work, but I feel they are being overly "clever" for no real reason, and make it harder to understand the spec. It would also be an unnecessary deviation from the official spec, which simply lists all the conversion instruction under *instr*, instead of grouping them with the unary and binary operators (page 11 in the current pdf).

In the vein of "keep it simple, stupid", I'm leaning towards letting the Conversions section be a self-contained machinery that can deal with it's domain problems in its own way.

ConvOp
--------
Another, more extensible approach would be to group conversion operators under `ConvOp` types that would handle correct pushing and popping with type checks, much like `UnOp` and `BinOp`. The fact that sign information comes last makes it harder to make a neat syntax though. Here's something I tried, but it broke the use of `_` as a wildcard in rules, since it becomes a token, which seems like a high price to pay.

```k
    syntax Instr ::= ValType "." ConvOp "_" ValType
                   | ValType "." SConvOp "_" ValType "_" Sx
                   | #conv (     ValType, ConvOp,  Number )
                   | #sconv( Sx, ValType, SConvOp, Number )

 // -------------------------------------------------------

    syntax Sx ::= "u" | "s"
 // -----------------------
    rule <k> TARGTYPE . COP _ SRCTYPE => #conv(TARGTYPE, COP, N) ... </k>
         <stack> < SRCTYPE > N : STACK => STACK </stack>

    rule <k> TARGTYPE . COP _ SRCTYPE _ S => #sconv(S, TARGTYPE, COP, N) ... </k>
         <stack> < SRCTYPE > N : STACK => STACK </stack>



    syntax ConvOp ::= "wrap"
 // ------------------------
    rule <k> #conv(i32, wrap, I) => #chop(< i32 > I) ... </k>
```

Conclusion
-------------
I think the solution in the commit seems easiest to read, and keeps conversions separate from other instructions. If it causes problems down the road, I'm happy to adjust course.

Reply:
These are clever, but I think I agree that the current solution is the easiest to read. Thanks for investigating!